@ortigali This is a backport of your PR. It failed PR validation. I guess same thing can happen in master. Could you look at this failure?


```
[info] - must be able to persist events that happen during recovery *** FAILED *** (12 milliseconds)
[info]   java.lang.AssertionError: assertion failed: found unexpected List(a-1, a-2)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsgAnyOf_internal(TestKit.scala:494)
[info]   at akka.testkit.TestKitBase$class.expectMsgAnyOf(TestKit.scala:480)
[info]   at akka.testkit.TestKit.expectMsgAnyOf(TestKit.scala:814)
[info]   at akka.persistence.PersistentActorSpec$$anonfun$1$$anonfun$apply$mcV$sp$44.apply(PersistentActorSpec.scala:1157)
[info]   at akka.persistence.PersistentActorSpec$$anonfun$1$$anonfun$apply$mcV$sp$44.apply(PersistentActorSpec.scala:1154)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
```

--- a/akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
+++ b/akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala
@@ -318,6 +318,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param handler handler for each persisted `event`
    */
   def persist[A](event: A)(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.")
     pendingStashingPersistInvocations += 1
     pendingInvocations addLast StashingHandlerInvocation(event, handler.asInstanceOf[Any ⇒ Unit])
     eventBatch ::= AtomicWrite(PersistentRepr(event, persistenceId = persistenceId,
@@ -333,6 +334,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param handler handler for each persisted `events`
    */
   def persistAll[A](events: immutable.Seq[A])(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.")
     if (events.nonEmpty) {
       events.foreach { event ⇒
         pendingStashingPersistInvocations += 1
@@ -371,6 +373,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param handler handler for each persisted `event`
    */
   def persistAsync[A](event: A)(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.")
     pendingInvocations addLast AsyncHandlerInvocation(event, handler.asInstanceOf[Any ⇒ Unit])
     eventBatch ::= AtomicWrite(PersistentRepr(event, persistenceId = persistenceId,
       sequenceNr = nextSequenceNr(), writerUuid = writerUuid, sender = sender()))
@@ -384,7 +387,8 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param events events to be persisted
    * @param handler handler for each persisted `events`
    */
-  def persistAllAsync[A](events: immutable.Seq[A])(handler: A ⇒ Unit): Unit =
+  def persistAllAsync[A](events: immutable.Seq[A])(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.")
     if (events.nonEmpty) {
       events.foreach { event ⇒
         pendingInvocations addLast AsyncHandlerInvocation(event, handler.asInstanceOf[Any ⇒ Unit])
@@ -392,6 +396,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
       eventBatch ::= AtomicWrite(events.map(PersistentRepr(_, persistenceId = persistenceId,
         sequenceNr = nextSequenceNr(), writerUuid = writerUuid, sender = sender())))
     }
+  }
 
   @deprecated("use persistAllAsync instead", "2.4")
   def persistAsync[A](events: immutable.Seq[A])(handler: A ⇒ Unit): Unit =
@@ -415,6 +420,7 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    * @param handler handler for the given `event`
    */
   def deferAsync[A](event: A)(handler: A ⇒ Unit): Unit = {
+    if (recoveryRunning) throw new IllegalStateException("Cannot persist during replay. Events can be persisted when receiving RecoveryCompleted or later.")
     if (pendingInvocations.isEmpty) {
       handler(event)
     } else {
@@ -539,10 +545,11 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
         context.system.scheduler.schedule(timeout, timeout, self, RecoveryTick(snapshot = false))
       }
       var eventSeenInInterval = false
+      var _recoveryRunning = true
 
       override def toString: String = "replay started"
 
-      override def recoveryRunning: Boolean = true
+      override def recoveryRunning: Boolean = _recoveryRunning
 
       override def stateReceive(receive: Receive, message: Any) = message match {
         case ReplayedMessage(p) ⇒
@@ -558,18 +565,18 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
         case RecoverySuccess(highestSeqNr) ⇒
           timeoutCancellable.cancel()
           onReplaySuccess() // callback for subclass implementation
-          changeState(processingCommands)
           sequenceNr = highestSeqNr
           setLastSequenceNr(highestSeqNr)
-          internalStash.unstashAll()
-          Eventsourced.super.aroundReceive(recoveryBehavior, RecoveryCompleted)
+          _recoveryRunning = false
+          try Eventsourced.super.aroundReceive(recoveryBehavior, RecoveryCompleted)
+          finally transitToProcessingState()
         case ReplayMessagesFailure(cause) ⇒
           timeoutCancellable.cancel()
           try onRecoveryFailure(cause, event = None) finally context.stop(self)
         case RecoveryTick(false) if !eventSeenInInterval ⇒
           timeoutCancellable.cancel()
           try onRecoveryFailure(
-            new RecoveryTimedOut(s"Recovery timed out, didn't get event within $timeout, highest sequence number seen $sequenceNr"),
+            new RecoveryTimedOut(s"Recovery timed out, didn't get event within $timeout, highest sequence number seen $lastSequenceNr"),
             event = None)
           finally context.stop(self)
         case RecoveryTick(false) ⇒
@@ -578,6 +585,17 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
         // snapshot tick, ignore
         case other ⇒
           stashInternally(other)
+      }
+
+      private def transitToProcessingState(): Unit = {
+        if (eventBatch.nonEmpty) flushBatch()
+
+        if (pendingStashingPersistInvocations > 0) changeState(persistingEvents)
+        else {
+          changeState(processingCommands)
+          internalStash.unstashAll()
+        }
+
       }
     }
 

--- a/akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala
+++ b/akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala
@@ -43,7 +43,7 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
       case "fail"                  ⇒ ReplayFilter.Fail
       case "warn"                  ⇒ ReplayFilter.Warn
       case other ⇒ throw new IllegalArgumentException(
-        s"invalid replay-filter.mode [$other], supported values [off, repair, fail, warn]")
+        s"invalid replay-filter.mode [$other], supported values [off, repair-by-discard-old, fail, warn]")
     }
   private def isReplayFilterEnabled: Boolean = replayFilterMode != ReplayFilter.Disabled
   private val replayFilterWindowSize: Int = config.getInt("replay-filter.window-size")

--- a/akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala
+++ b/akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala
@@ -1154,9 +1154,7 @@ abstract class PersistentActorSpec(config: Config) extends PersistenceSpec(confi
     "be able to persist events that happen during recovery" in {
       val persistentActor = namedPersistentActor[PersistInRecovery]
       persistentActor ! GetState
-      expectMsg(List("a-1", "a-2", "rc-1", "rc-2"))
-      persistentActor ! GetState
-      expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3"))
+      expectMsgAnyOf(List("a-1", "a-2", "rc-1", "rc-2"), List("a-1", "a-2", "rc-1", "rc-2", "rc-3"))
       persistentActor ! Cmd("invalid")
       persistentActor ! GetState
       expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3", "invalid"))

--- a/akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala
+++ b/akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala
@@ -646,6 +646,23 @@ object PersistentActorSpec {
     }
   }
 
+  class PersistInRecovery(name: String) extends ExamplePersistentActor(name) {
+    override def receiveRecover = {
+      case Evt("invalid") ⇒
+        persist(Evt("invalid-recovery"))(updateState)
+      case e: Evt ⇒ updateState(e)
+      case RecoveryCompleted ⇒
+        persistAsync(Evt("rc-1"))(updateState)
+        persist(Evt("rc-2"))(updateState)
+        persistAsync(Evt("rc-3"))(updateState)
+    }
+
+    override def onRecoveryFailure(cause: scala.Throwable, event: Option[Any]): Unit = ()
+
+    def receiveCommand = commonBehavior orElse {
+      case Cmd(d) ⇒ persist(Evt(d))(updateState)
+    }
+  }
 }
 
 abstract class PersistentActorSpec(config: Config) extends PersistenceSpec(config) with ImplicitSender {
@@ -1132,6 +1149,20 @@ abstract class PersistentActorSpec(config: Config) extends PersistenceSpec(confi
       val persistentActor = namedPersistentActor[RecoverMessageCausedRestart]
       persistentActor ! "Boom"
       expectMsg("failed with TestException while processing Boom")
+    }
+
+    "be able to persist events that happen during recovery" in {
+      val persistentActor = namedPersistentActor[PersistInRecovery]
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2"))
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3"))
+      persistentActor ! Cmd("invalid")
+      persistentActor ! GetState
+      expectMsg(List("a-1", "a-2", "rc-1", "rc-2", "rc-3", "invalid"))
+      watch(persistentActor)
+      persistentActor ! "boom"
+      expectTerminated(persistentActor)
     }
   }
 

